src/injectors-cache.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * Contains functions that inject dynamic content.
+ */
+
+import * as brotli from './brotli-wasm-wrapper';
+import {
+  ContentEncoding,
+  HeaderKeys,
+  IncomingCloudflarePropertiesExtended,
+  supportsBrotli,
+} from './headers';
+
+const INJECTORS_CACHE_NAME = 'injectors';
+const textEncoder = new TextEncoder();
+
+let injectorsCache_: Cache;
+/** */
+async function openInjectorsCache(): Promise<Cache> {
+  if (!injectorsCache_) {
+    injectorsCache_ = await caches.open(INJECTORS_CACHE_NAME);
+  }
+  return injectorsCache_;
+}
+
+/**
+ * Generates a complete cache key for a cache object.
+ *
+ * @param parts - unique list of identifiers for object in cache.
+ * @returns a unique cache key.
+ */
+function cacheKey(...parts: string[]): string {
+  return parts.join(';');
+}
+
+/**
+ * Gets an already injected Response for the given URL.
+ *
+ * @param url - base file URL that has dynamic content injected into it.
+ * @param key - unique identifiers for this specific injection.
+ * @param cf - incoming Cloudflare properties, used to determine whether to try
+ *     fetching the Brotli pre-compressed version of the URL.
+ * @returns an already injected Response.
+ */
+export async function getCacheFor(
+  url: string,
+  key: string,
+  cf: IncomingCloudflarePropertiesExtended
+): Promise<Response | undefined> {
+  const cache = await openInjectorsCache();
+  const keyParts = [key];
+  if (supportsBrotli(cf)) {
+    keyParts.push(ContentEncoding.BROTLI);
+  }
+  return cache.match(cacheKey(url, ...keyParts));
+}
+
+/**
+ * Saves an injected Response to the cache twice, for raw and Brotli-compressed.
+ *
+ * @param response - raw Response to save to cache.
+ * @param url - base file URL that has dynamic content injected into it.
+ * @param key - unique identifiers for this specific injection.
+ */
+export async function saveCache(
+  response: Response,
+  url: string,
+  key: string
+): Promise<void> {
+  if (!response.body) {
+    throw new Error('Response has no body');
+  }
+
+  // Tee the body as it is going to be read twice - once for saving the plain
+  // response and once for performing Brotli compression.
+  const [body1, body2] = response.body.tee();
+  const plainResponse = new Response(body1, response);
+  const bodyAsText = await new Response(body2, response).text();
+
+  // Create an equivalent Response object that is already Brotli-compressed.
+  const compressed = await brotli.compress(textEncoder.encode(bodyAsText), {
```

Comment:
Could we avoid the new response, decode `.text()`, and encode by using `body2.arrayBuffer()`?

Reply:
Unfortunately no, `body2` is a `ReadableStream<Uint8Array>`, it doesn't have an `arrayBuffer` method - I tried playing around with the various methods but couldn't find an easy way to convert the `ReadableStream<Uint8Array>` to a `Uint8Array` (!) which is what `compress` takes. Maybe you have a better idea!

Reply:
I didn't realize it was a `ReadableStream<Uint8Array>`, I figured it was already a `Uint8Array`. In that case, would `new Response(body2).arrayBuffer()` give us a proper `ArrayBuffer`, which we can pass to `new Uint8Array(buffer)`? I know the `new Uint8Array` will use the already allocated buffer, and I'm hoping the `.arrayBuffer()` will be able to peek a hole into the already allocated internal array buffer.

We can further use `response.clone()` to create the second response to read the array buffer.

Reply:
As I replied earlier, we can only go through a `response`'s body (whether we read it using `.text()`, `.json()`, `.blob()`, etc...) once, so we have to use the `.tee`'ed version to pass it to the compression

Honestly how come there's no `Uint8Array.from(<ReadableStream<Uint8Array>>)`? :(

Reply:
Okay, so I did some stackoverflow deep diving and landed on this change: 4c79f34
Still requires a transitional response object, but skips the TextEncoder step (thankfully!)